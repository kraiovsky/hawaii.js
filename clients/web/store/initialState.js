export default {
  isFinished: {
    auth: false,
  },
  auth: {
    access_token: null,
    refresh_token: null,
    scope: null,
    uid: null,
    email: null,
  },
  pageTitle: '',
}
