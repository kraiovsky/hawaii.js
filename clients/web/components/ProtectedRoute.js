import React from 'react'
import Link from 'next/link'
import { useSelector } from 'react-redux'
import PropTypes from 'prop-types'
import { getIsAuthorizedUser, getIsFinished } from '../store'

const ProtectedRoute = ({ scope = [], children }) => {
  const state = useSelector(state => state)
  let content = <></>
  if (getIsFinished(state, 'auth') && getIsAuthorizedUser(state, scope)) {
    content = <div>{children}</div>
  }
  if (getIsFinished(state, 'auth') && !getIsAuthorizedUser(state, scope)) {
    content = (
      <p>
        You need to{' '}
        <Link href="/auth">
          <a>login</a>
        </Link>{' '}
        to access this page.
      </p>
    )
  }
  return content
}

ProtectedRoute.propTypes = {
  scope: PropTypes.arrayOf(PropTypes.string).isRequired,
  children: PropTypes.element.isRequired,
}

export default ProtectedRoute
