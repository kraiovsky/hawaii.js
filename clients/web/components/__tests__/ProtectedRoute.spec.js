import React from 'react'
import renderer from 'react-test-renderer'
import { useSelector } from 'react-redux'
import initialState from '../../store/initialState'
import ProtectedRoute from '../ProtectedRoute'

jest.mock('react-redux')
const scope = ['admin']
const children = <>Protected Route Content</>

describe('Render ProtectedRoute', () => {
  test('👎 Unauthenticated user', () => {
    useSelector.mockImplementation(() => ({
      ...initialState,
      isFinished: {
        auth: true,
      },
    }))
    const component = renderer.create(<ProtectedRoute scope={scope} children={children} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('👎 Authenticated but not authorized user', () => {
    useSelector.mockImplementation(() => ({
      ...initialState,
      isFinished: {
        auth: true,
      },
      auth: {
        access_token: 'some-access-token',
        refresh_token: 'some-refresh-token',
        scope: 'super-user',
        uid: '123-user',
        email: 'user@email.com',
      },
    }))
    const component = renderer.create(<ProtectedRoute scope={scope} children={children} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('👍 Authenticated and authorized user', () => {
    useSelector.mockImplementation(() => ({
      ...initialState,
      isFinished: {
        auth: true,
      },
      auth: {
        access_token: 'some-access-token',
        refresh_token: 'some-refresh-token',
        scope: 'admin',
        uid: '123-user',
        email: 'user@email.com',
      },
    }))
    const component = renderer.create(<ProtectedRoute scope={scope} children={children} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
