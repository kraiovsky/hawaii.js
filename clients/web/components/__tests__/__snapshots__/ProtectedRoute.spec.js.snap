// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ProtectedRoute 👍 Authenticated and authorized user 1`] = `
<div>
  Protected Route Content
</div>
`;

exports[`Render ProtectedRoute 👎 Authenticated but not authorized user 1`] = `
<p>
  You need to
   
  <a
    href="/auth"
    onClick={[Function]}
  >
    login
  </a>
   
  to access this page.
</p>
`;

exports[`Render ProtectedRoute 👎 Unauthenticated user 1`] = `
<p>
  You need to
   
  <a
    href="/auth"
    onClick={[Function]}
  >
    login
  </a>
   
  to access this page.
</p>
`;
